compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1707,84 +1754,27 @@ object Types {
-           |long form = $toString of class $getClass,
+          )
+        ||
+        sym == defn.AnyClass.primaryConstructor,
+        s"""data race? overwriting $lastSymbol with $sym in type $this,
            |last sym id = ${lastSymbol.id}, new sym id = ${sym.id},
            |last owner = ${lastSymbol.owner}, new owner = ${sym.owner},
            |period = ${ctx.phase} at run ${ctx.runId}""")
     }
 
-    private[dotc] def withDenot(denot: Denotation)(implicit ctx: Context): ThisType = {
-      val adapted =
-      	if (hasFixedSym)
-      	  this
-      	else if (signature != denot.signature)
-          withSig(denot.signature)
-        else if (denot.symbol.isPrivate)
-          withNameSpace(denot.symbol.owner.typeRef)
-        else
-          this
-      if (adapted ne this) adapted.withDenot(denot).asInstanceOf[ThisType]
-      else {
-        setDenot(denot)
-        this
-      }
-    }
-
-    private[dotc] final def setDenot(denot: Denotation)(implicit ctx: Context): Unit = {
-      if (Config.checkNoDoubleBindings)
-        if (ctx.settings.YnoDoubleBindings.value)
-          checkSymAssign(denot.symbol)
-
-      // additional checks that intercept `denot` can be added here
-
-      lastDenotation = denot
-      checkDenot()
-      lastSymbol = denot.symbol
-      checkedPeriod = Nowhere
-    }
-
-    private[dotc] def withSym(sym: Symbol)(implicit ctx: Context): this.type = {
-      setSym(sym)
-      this
-    }
-
-    private[dotc] final def setSym(sym: Symbol)(implicit ctx: Context): Unit = {
-      if (Config.checkNoDoubleBindings)
-        if (ctx.settings.YnoDoubleBindings.value)
-          checkSymAssign(sym)
-      uncheckedSetSym(sym)
-    }
-
-    private[dotc] final def uncheckedSetSym(sym: Symbol): Unit = {
-      lastDenotation = null
-      lastSymbol = sym
-      checkedPeriod = Nowhere
-    }
-
-    private def withSig(sig: Signature)(implicit ctx: Context): NamedType =
-      TermRef(prefix, designator.withSig(sig))
+    /** A reference with the initial symbol in `symd` has an info that
+     *  might depend on the given prefix.
+     */
+    private def infoDependsOnPrefix(symd: SymDenotation, prefix: Type)(implicit ctx: Context): Boolean =
+      symd.maybeOwner.membersNeedAsSeenFrom(prefix) && !symd.is(NonMember)
 
-    protected def loadDenot(implicit ctx: Context): Denotation = {
-      val d = asMemberOf(prefix, allowPrivate = false)
-      if (d.exists || ctx.phaseId == FirstPhaseId || !lastDenotation.isInstanceOf[SymDenotation])
-        if (mySig != null && mySig.ne(Signature.OverloadedSignature)) d.atSignature(mySig).checkUnique
-        else d
-      else { // name has changed; try load in earlier phase and make current
-        val d = loadDenot(ctx.withPhase(ctx.phaseId - 1)).current
-        if (d.exists) d
-        else throw new Error(s"failure to reload $this of class $getClass")
-      }
+    /** Is this a reference to a class or object member? */
```

Comment:
This comment does not seem to match the implementation: `infoDependsOnPrefix` calls `membersNeedAsSeenFrom(prefix)` which will return false if `prefix` is a statically accessible object.